pageserver/src/layered_repository/timeline.rs

```diff
@@ -1580,58 +1579,73 @@ impl LayeredTimeline {
 
         let first_level0_delta = level0_deltas_iter.next().unwrap();
         let mut prev_lsn_end = first_level0_delta.get_lsn_range().end;
-        let mut deltas_to_compact = vec![Arc::clone(first_level0_delta)];
+        let mut deltas_to_compact = vec![first_level0_delta];
         for l in level0_deltas_iter {
             let lsn_range = l.get_lsn_range();
 
             if lsn_range.start != prev_lsn_end {
                 break;
             }
-            deltas_to_compact.push(Arc::clone(l));
+            deltas_to_compact.push(l);
             prev_lsn_end = lsn_range.end;
         }
         let lsn_range = Range {
             start: deltas_to_compact.first().unwrap().get_lsn_range().start,
             end: deltas_to_compact.last().unwrap().get_lsn_range().end,
         };
 
+        let deltas_to_compact_count = deltas_to_compact.len();
         info!(
             "Starting Level0 compaction in LSN range {}-{} for {} layers ({} deltas in total)",
             lsn_range.start,
             lsn_range.end,
-            deltas_to_compact.len(),
+            deltas_to_compact_count,
             level0_deltas.len()
         );
         for l in deltas_to_compact.iter() {
             info!("compact includes {}", l.filename().display());
         }
-        // We don't need the original list of layers anymore. Drop it so that
-        // we don't accidentally use it later in the function.
-        drop(level0_deltas);
 
         // This iterator walks through all key-value pairs from all the layers
         // we're compacting, in key, LSN order.
-        let all_values_iter = deltas_to_compact
-            .iter()
-            .map(|l| l.iter())
-            .kmerge_by(|a, b| {
-                if let Ok((a_key, a_lsn, _)) = a {
-                    if let Ok((b_key, b_lsn, _)) = b {
-                        match a_key.cmp(b_key) {
-                            Ordering::Less => true,
-                            Ordering::Equal => a_lsn <= b_lsn,
-                            Ordering::Greater => false,
-                        }
-                    } else {
-                        false
+        let mut lsn_range = None::<Range<Lsn>>;
+        let mut all_values_vec: Vec<anyhow::Result<(Key, Lsn, Value)>> = Vec::new();
+        let mut level0_delta_files_to_remove = HashSet::with_capacity(level0_deltas.len());
+
+        for l in &level0_deltas {
+            lsn_range = match lsn_range {
+                Some(old_range) => {
+                    let new_range = l.get_lsn_range();
+                    Some(min(old_range.start, new_range.start)..max(old_range.end, new_range.end))
+                }
+                None => Some(l.get_lsn_range()),
+            };
+
+            all_values_vec.extend(l.iter());
+            level0_delta_files_to_remove.insert(l.path());
+        }
```

Comment:
This collects all keys and all values from all of the delta layers into one giant Vec. That's not cool, each delta layer file is around 200-1000 MB in size so this consumes several gigabytes of memory. Let's do the merging "on the fly" like before.

Later in the function, we collect a Vec with all Keys, which also takes quite a lot of memory and that's not great either. Would be nice to avoid that too, but it's not as bad as collecting all keys and values.

Reply:
Would be good to add a comment on that while we're at it. It's not obvious if you don't realize that the iterators return a lot of data.